toolchain/lex/string_literal.cpp

```diff
@@ -466,21 +476,82 @@ static auto ExpandEscapeSequencesAndRemoveIndent(
   }
 }
 
-auto StringLiteral::ComputeValue(llvm::BumpPtrAllocator& allocator,
-                                 DiagnosticEmitter& emitter) const
-    -> llvm::StringRef {
-  if (!is_terminated_) {
-    return "";
+auto StringLiteral::ComputeCharValue(Diagnostics::Emitter<const char*>& emitter)
+    const -> std::optional<CharLiteralValue> {
+  CARBON_DCHECK(kind_ == Kind::Char);
+  CARBON_DCHECK(is_terminated_);
+
+  if (hash_level_ != 0) {
+    CARBON_DIAGNOSTIC(CharLiteralRaw, Error,
+                      "unexpected `#` before character literal");
+    emitter.Emit(text_.begin(), CharLiteralRaw);
   }
-  if (multi_line_ == MultiLineWithDoubleQuotes) {
+
+  // Allocate a buffer sized to the content. Note it's possible this could be
+  // more efficient/faster with a `ExpandEscapeSequencesAndRemoveIndent`
+  // implementation aware of the buffer size, but this is trying to share logic
+  // with string expansion.
+  llvm::SmallVector<char, 0> buffer;
+  buffer.resize_for_overwrite(content_.size());
```

Comment:
This always performs a heap allocation, which seems unnecessary. All valid cases should fit in a buffer of only a few bytes. Maybe we could use the default `SmallVector` buffer size instead?

(Also would it make sense to use an `llvm::SmallString` given that this is a string buffer?)